### ISSUE TITLE
[ENGINEERING] LCX retrieval: replace torch.topk with fused double-argmax (free 1.2ms/step)

### ISSUE THREAD

**Issue by Kenessy** [VRAXION/VRAXION#95]
## Summary

Outcome of #91 (top_k sweep) + deep research into retrieval kernel costs.

**Finding:** At k=2 and ~500 bucket candidates, the `torch.topk` call contributes ~300µs of Python dispatch overhead **per call site**. There are 4 call sites in `swarm_model.py`. The actual compute (finding top-2 from 500 floats) costs <200ns. Framework overhead is **1500× the work itself.**

Two `.argmax()` calls are semantically identical to `topk(k=2)` but bypass the topk dispatch path entirely.

---

## Scope

**File:** `S:/AI/work/VRAXION_DEV/Diamond Code/swarm_model.py`

**4 call sites to replace:**

| Line | Context | Shape | Notes |
|---|---|---|---|
| 2337 | Read, small level (no bucket) | `[B, S]` | Primary read path |
| 2373 | Read, large level (bucketed) | `[B, max_bkt]` | Hot path at 2000+ slots |
| 2390 | Score margin diagnostic | `[B, S/max_bkt]` | Already in `no_grad` |
| 2433 | Write path | `[B, S]` | Already in `no_grad` |

---

## Implementation

Replace each `scores.topk(eff_k, dim=-1)` where `eff_k=2` with:

```python
# Two-pass double argmax — O(n) exact, no partial sort, no dispatch overhead
i1 = scores.argmax(dim=-1, keepdim=True)                      # [B, 1]
s2 = scores.scatter(-1, i1, -float('inf'))                    # mask winner
i2 = s2.argmax(dim=-1, keepdim=True)                          # [B, 1]
topk_idx    = torch.cat([i1, i2], dim=-1)                     # [B, 2]
topk_scores = scores.gather(-1, topk_idx)                     # [B, 2]
```

This is **branch-free, SIMD-vectorized by PyTorch automatically, and produces identical output** to `topk(k=2)`.

---

## Guard

Add a config check — if `lcx_top_k > 2`, fall back to `torch.topk` (keep backward compat):

```python
if self.lcx_top_k == 2:
    # fast path
    ...
else:
    topk_scores, topk_idx = scores.topk(eff_k, dim=-1)
```

---

## Expected impact

- **4 call sites × ~300µs saved = ~1.2ms per forward pass** on CPU
- At D=6180 on GPU: dispatch overhead smaller but still measurable (~50µs/call)
- Zero accuracy change — mathematically identical
- Zero risk — trivial swap, easy to revert

---

## Probe required?

No — this is pure equivalence refactor. Run existing mini-model smoke test to confirm:
`python tools/_scratch/test_c19_bn_mini.py` before + after, verify bit_acc identical.

---

## Connection

- Closes after: #91 (top_k locked at 2)
- Precedes: #96 (PQ scoring — the real bottleneck)
- Research basis: PyTorch issue #8817, NumPy issue #25846 (dispatch overhead measured)

**Comment by Kenessy** [comment]
## Probe result: REJECTED — topk is 4× faster than double-argmax on CPU

Ran `probe_argmax_sanity.py` (2026-02-19). Results:

### Timing (5000 iterations each, CPU fp32)

| Shape | topk(2) | argmax×2 | winner |
|---|---|---|---|
| [32 × 512]  | **76µs**  | 306µs | topk 4.0× faster |
| [32 × 2000] | **153µs** | 583µs | topk 3.8× faster |
| [10 × 512]  | **30µs**  | 113µs | topk 3.8× faster |
| [1  × 512]  | **7µs**   | 30µs  | topk 4.4× faster |

### Why the theory was wrong

`torch.topk` is a **single fused C++ kernel** — one Python dispatch call, one SIMD pass.

Double-argmax requires **4 separate Python dispatch calls**: `argmax → scatter → argmax → cat`. Each carries its own overhead. The overhead multiplies, not cancels.

The benchmark numbers from the research (~300µs per call) assumed each operation independently — real measured cost shows topk bundles everything in one trip at ~76µs.

### Unit correctness

Unit test PASSED: 10,000 random tensors, zero mismatches between topk and argmax outputs.

Integration test PASSED: bit_acc=0.753 after 100 steps, score margin firing 100/100 steps.

### Action

- Reverted all 3 swapped sites in swarm_model.py back to `scores.topk(eff_k, dim=-1)`
- `probe_argmax_sanity.py` kept in _scratch/ as regression test
- **Closing this ticket** — topk is already optimal for CPU. Effort should go to #96 (PQ scoring) which targets the dot-product computation, not the selection step.
